[LiveComponent] Support for liveProp union type

kevinmade opened the pull request:
| Q              | A
| -------------- | ---
| Bug fix?       | no
| New feature?   | yes
| Deprecations?  | no
| Documentation? | no
| Issues         | Fix #853
| License        | MIT

This PR adds support for liveProp union types in Live Components. Previously, using a union type would throw an exception. With this change `LiveComponentMetadataFactory` will be able to create liveProp metadata for union types, but will only work in combination with a custom hydrate and dehydrate implementation.

raphael-geffroy committed:
[Autocomplete] Add documentation about plugins management

Kocal committed:
[Site] Document Toolkit recipes props and blocks

bernard-ng committed:
[Toolkit][Shadcn] Add Item component

kevinmade committed:
Fix liveProp union type

Kocal reviewed (changes requested):
Hi and thanks for the contribution!

This new feature needs to be covered with some tests, and the CHANGELOG.md must be updated too.

Also, you only support the TypeInfo path and not the legacy PropertyInfo path, but I think that's fine.

kevinmade committed:
Update changelog

kevinmade committed:
Update changelog

kevinmade committed:
Add LiveComponentMetadataFactory union type test

kevinmade committed:
Fix LiveComponentMetadataFactoryTest testLivePropUnionType

kevinmade committed:
Fix LiveComponentMetadataFactoryTest testLivePropUnionType

kevinmade committed:
Fix php-cs-fixer

alinceDev committed:
[TwigComponent][Perf] Add option to disable the dump of components

Commit:
[Autocomplete] Fix issue where TomSelect could already been initialized on an element

ker0x committed:
[Toolkit][Shadcn] Update Button CSS class, remove as props

Kocal committed:
[CI] Tests over Symfony 7.4.0-beta1 and 8.0.0-beta1

bernard-ng committed:
[Toolkit][Shadcn] docs for Item component

Kocal committed:
[Toolkit][Shadcn][Item] Minor fixes (html_cva, missing dependencies, more examples)

Kocal committed:
[Toolkit][Shadcn][Button] Re-introduce `as` prop, adjust some examples

Kocal committed:
[Toolkit] Add CONTRIBUTING.md

Wertisdk committed:
Add clarification about difference between component re-rendering and initial loading

Add a note that explains that the re-rendering loading states is a different concept from the components initial loading state regarding placeholders, with a link to the related "loading content" section in this doc

ker0x committed:
[Toolkit][Shadcn] Add missing as props, update CONTRIBUTING.md

[Toolkit][Shadcn] Add missing as props, update CONTRIBUTING.md

Kocal committed:
[TwigComponent] Merge `profiler` and `profiler_collect_components` config options

alvidias committed:
fix: add missing closing quote to attribute

andyexeter committed:
Allow using directory name as component name for anonymous components

smnandre commented on src/LiveComponent/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+- Add support for `LiveProp` union types in Live Components.
+  `LiveComponentMetadataFactory` can now create `liveProp` metadata for union types.
+  Note: This feature only works when using a custom `hydrate` and `dehydrate` implementation.
```
Hey @kevinmade ! Could you add a test or two covering the cases "not supposed to work" (ie: when hydrate and/or dehydrate are missing ?)

kevinmade replied:
I thought a hydrator and dehydrator were required for my change, but this had something to do with my own use case, where the union had a type that wasn't a hydratable. I’ve removed the note from the changelog and added a test.

seb-jean committed:
[Toolkit][Shadcn] Add Dialog recipe

kevinmade committed:
Add testObjectTypedLivePropCannotBeDehydrated and update CHANGELOG.md

Kocal committed:
Fix CI when testing Symfony RCs versions

Kocal committed:
[Notify][Turbo] Add support for MercureBundle ^0.4.0 and Mercure ^0.7.0

Kocal committed:
[Toolkit][Shadcn] Minor tweaks on alert-dialog and dialog cmoponents

Kocal committed:
[Toolkit] Pin `tales-from-a-dev/twig-tailwind-extra` to `^1.0.0`

Kocal committed:
Tests on Symfony 8 stable

Commit:
[Toolkit][Kdb] Add new component kdb from shadcn ui

tassilogroeper committed:
bug symfony#3197 [Toolkit][Badge] Remove hover style to clearly distinguish from Button (Tassilo Gröper)

This PR was merged into the 2.x branch.

Discussion
----------

[Toolkit][Badge] Remove hover style to clearly distinguish from Button

| Q              | A
| -------------- | ---
| Bug fix?       | yes
| New feature?   | no
| Deprecations?  | no
| Documentation? | no
| Issues         |
| License        | MIT

Confusion may occur, if the badge shows a different style on mouse hover. As it is not a button, we need to remove it.

Additionally, the badge component on Shadcn does not hold hover styles, when interacted with a mouse. see  https://ui.shadcn.com/docs/components/badge

Commits
-------

f14635b [Toolkit][Badge] Remove hover style to clearly distinguish badge from button

Kocal committed:
[Icons] Fix `aliases` and `icon_attributes` config types for `config/reference.php`

Kocal committed:
Git-ignore config/reference.php

Kocal committed:
[Translator] Refactor API to use string-based translation keys instead of generated constants

Kocal committed:
Update E2E app

kasperh90 committed:
docs: fix missing space in JSON snippet for UX Autocomplete

Kocal committed:
Update ux.symfony.com app

Kocal committed:
Update Encore app

bernard-ng committed:
[Toolkit][Shadcn] add Empty component

bernard-ng committed:
[Toolkit][Shadcn] add Field component

bernard-ng committed:
[Toolkit][Shadcn] add ButtonGroup component

Kocal committed:
[Translator] Add option `ux_translator.dump_typescript` to enable/disable TypeScript types generation

Kocal committed:
[Translator] Early exit parameters extraction from Intl messages, if no `{` is found

smnandre committed:
[minor] fix upson logo & url

Kocal committed:
[Translator] Reword `dump_typescript` option description

Kocal committed:
[Site] Hide Toolkit menu

Kocal commented on src/LiveComponent/tests/Integration/LiveComponentHydratorTest.php:
```diff
@@ -1636,6 +1636,23 @@ public function testInterfaceTypedLivePropCannotBeDehydrated()
         $this->hydrator()->dehydrate($component, $this->createComponentAttributes(), $this->createLiveMetadata($component));
     }
 
+    public function testObjectTypedLivePropCannotBeDehydrated()
+    {
+        $componentClass = new class {
+            #[LiveProp(writable: true)]
+            public object $object;
+        };
+
+        $component = new $componentClass();
+        $component->object = new class {
+        };
+
+        $this->expectException(\LogicException::class);
+        $this->expectExceptionMessageMatches('/The "object" property on component ".*" is missing its property-type. Add the ".*" type so the object can be hydrated later./');
+
+        $this->hydrator()->dehydrate($component, $this->createComponentAttributes(), $this->createLiveMetadata($component));
+    }
```
To move in `\Symfony\UX\LiveComponent\Tests\Integration\LiveComponentHydratorTest::provideDehydrationHydrationTests`

Kocal commented on src/LiveComponent/tests/Functional/Metadata/LiveComponentMetadataFactoryTest.php:
```diff
@@ -42,4 +43,21 @@ public function testQueryStringMapping()
         $this->assertNotNull($propsMetadataByName['boundPropWithCustomAlias']);
     }
+
+    public function testLivePropUnionType()
+    {
+        /** @var LiveComponentMetadataFactory $metadataFactory */
+        $metadataFactory = self::getContainer()->get('ux.live_component.metadata_factory');
+
+        $class = new \ReflectionClass(WithUnionIntersectionType::class);
+        $propsMetadata = $metadataFactory->createPropMetadatas($class);
+
+        $propsMetadataByName = [];
+        foreach ($propsMetadata as $propMetadata) {
+            $propsMetadataByName[$propMetadata->getName()] = $propMetadata;
+        }
+
+        $this->assertEquals('mixed', (string) $propsMetadataByName['unionProp']->getType());
+        $this->assertEquals('mixed', (string) $propsMetadataByName['intersectionProp']->getType());
+    }
```
I don't think it makes sense to have this test

Kocal commented on src/LiveComponent/src/Metadata/LiveComponentMetadataFactory.php:
```diff
@@ -122,6 +123,10 @@ public function createLivePropMetadata(string $className, string $propertyName,
         } else {
+            if ($reflectionType instanceof \ReflectionUnionType || $reflectionType instanceof \ReflectionIntersectionType) {
+                return new LivePropMetadata($property->getName(), $liveProp, Type::mixed());
+            }
```
Why do we use `Type::mixed()` here, instead of `$this->typeResolver->resolve($property)` below?

Kocal reviewed (changes requested):
Please add tests in `LiveComponentHydratorTest::provideDehydrationHydrationTests` data provider, so we can test the full dehydration/hydration process. Testing only the metadata factory is not enough.

fabpot committed:
Add split.sh config

Kocal committed:
Add E2E tests for Autocomplete

Kocal committed:
[Site] Update UX packages

Kocal committed:
[Site] Fix examples from Translator page

Kocal committed:
[Site] Hide Toolkit from packages list

I prefer this solution over commenting Toolkit package from UxPackageRepository, because this way we can still secretly access the Toolkit-related pages.

Kocal committed:
Add E2E tests for LiveComponent

Commit:
feat(toolkit): add input-group component for shadcn kit

Kocal committed:
Add E2E tests for Chart.js

Kocal committed:
Add E2E tests for Turbo

Kocal committed:
Add E2E tests for CropperJS

Kocal committed:
[Translator] Refactor `TranslationsDumper` options from `__constructor` and setters, to `dump` method

StevenRenaux committed:
[Map] Add `Map::removeAll*()` methods

Kocal committed:
[Translator] Add `keys_patterns` configuration option to filter dumped translations by key patterns

Kocal committed:
[E2E] Add quality tools: PHP-CS-Fixer and Twig-CS-Fixer

Kocal committed:
Pin phpdocumentor/reflection-docblock to non dev version

Kocal committed:
Truly leverage PHP-CS-Fixer and Twig-CS-Fixer to Fabbot

Kocal committed:
[Pkg] Fix pnpm build command

Following https://github.com/symfony/ux/pull/2983 and https://github.com/symfony/ux/pull/3001, I noticed sometimes we still have the `pnpm run build` command that fails.

With a `sync` call and filtering out `@symfony/ux-map` when building other packages, I think we could finally fix the issue.

Kocal committed:
[Toolkit] Fix forgotten documentation links, follows #3097

Kocal committed:
[Site] Un-hide Toolkit, partially reverts #3224

Kocal committed:
[Site] Rework Toolkit recipe rendering system, improve our Markdown converter

Kocal committed:
[*CS-Fixer] Ensure "apps/" exists before adding it in Finder

Kocal committed:
[Toolkit][Shadcn] Update CSS file

JoppeDC committed:
Fix codeblock

ousmaneNdiaye committed:
Update index.rst with MarkdownTextarea template info

keradus committed:
PHP CS Fixer: update usage

Kocal committed:
Run Twig-CS-Fixer on all repo's files, follow #3263

smnandre committed:
[Site] Fix TerminalCommand layout

closes #3266

Kocal committed:
[Site] Wording on Shadcn installation

Kocal committed:
[Site] Update UX dependencies

Kocal committed:
[Site] Set dark/light theme of Toolkit preview iframes, given current page theme

GromNaN committed:
[Turbo] Fix SplObjectStorage::attach deprecation in Doctrine Broadcast listener

The method is deprecated in PHP 8.5

Also refactor to avoid dynamic property access and help static analysis

Kocal committed:
[CI] Fix installation of PropertyInfo and TypeInfo for LiveComponent

jeremyFreeAgent committed:
fix(tookit): fix getRecipes method with type

jeremyFreeAgent committed:
[Tookit] fix display of recipe name

Kocal committed:
[CI] Remove usage of framework.annotations.enabled

Done in https://github.com/symfony/symfony/pull/61174

Kocal committed:
[CI] Fix compatibility issues with Symfony 8 and spatie/phpunit-snapshot-assertions

ousmaneNdiaye committed:
Fix button syntax for file upload actions

Since self-closing the button tag is not recommended

jeremyFreeAgent committed:
[Toolkit] Prepare base for recipe type "Block"

Kocal committed:
[Toolkit][Shadcn] Add Tooltip component

Kocal committed:
[StimulusBundle][Performance] Change AssetMapper `excluded_patterns` from `**/controllers.json` to `*/controllers.json`

Related to https://github.com/symfony/symfony/issues/61771

Kocal committed:
Add changelog entry for #3285

Kocal committed:
[Site] Update to Symfony 7.4

Kocal committed:
[Site] Update to PHP 8.5

Kocal committed:
[Site] Update to Symfony 8

Kocal committed:
Update root JS dependencies

Jibbarth committed:
Fix text-color for destructive badge

Kocal committed:
[Site] Move website to dedicated repository

daFish committed:
[Toolkit][Shadcn] Add Accordion recipe

MrYamous committed:
remove installation note about assetmapper

Phil23 committed:
[TwigComponent] Add support for anonymous component name resolution in external bundle

Co-authored-by: Simon André <smn.andre@gmail.com>

Kocal committed:
[Toolkit][Shadcn] Fix text color in dark mode for AlertDialog/Dialog components

Cherry-picked from https://github.com/symfony/ux/pull/3299

Kocal committed:
Fix npm releases due to `repository` issue

Kocal committed:
Update versions to 2.32.0

Kocal committed:
[Toolkit][Shadcn] Rework many components examples

Kocal committed:
[Toolkit][Shadcn] Rework templates of `avatar` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `badge` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `card` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `checkbox` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `input-group` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `table` recipe

Kocal committed:
[Toolkit][Shadcn] Rework templates of `textarea` recipe

Kocal committed:
[Toolkit][Shadcn] Update snapshots

Kocal committed:
[React] Add support for React 19

Kocal committed:
[Toolkit][Shadcn] Fix `InputGroup` example

ker0x committed:
[Toolkit][Shadcn] Update Alert CSS class, add Alert:Action component

Kocal committed:
[Toolkit] Fix `Field` demo code

Kocal committed:
[Autocomplete][Chartjs][Cropperjs][Dropzone][LazyImage][React][StimulusBundle][Svelte][Swup][TogglePassword][Translator][Turbo][Typed][Vue] Use `Extension` from `DependencyInjection` instead of `HttpKernel`

seb-jean committed:
[Icons] Add support for suffixes

Kocal committed:
Fix minimum requested 6.x version of FrameworkBundle for LiveComponent

Related to https://github.com/symfony/ux/actions/runs/21540501201/job/62074166912?pr=3318#step:9:1262
AbstractBundle has been introduced in 6.1: https://github.com/symfony/symfony/commit/7e8cf5df105ab0bbe0112222fa5498af486d929c

ker0x committed:
[Toolkit][Shadcn] Update Avatar component

kkevindev committed:
Update GitHub link in documentation

Kocal committed:
[TwigComponent] Document high-order components

Kocal committed:
[TwigComponent] Document about Readonly  components

Kocal committed:
Run PHP-CS-Fixer (no_useless_else & static_lambda)

Kocal committed:
Drop Biome.js for oxfmt and oxlint

Kocal committed:
-

Kocal committed:
[LiveComponent] Add option `fetchCredentials`

GromNaN committed:
zenstruck/foundry bundle requires symfony/http-kernel 6.1

zenstruck/foundry removed the requirement for symfony/framework-bundle in version 5.0.

https://github.com/zenstruck/foundry/commit/0b09c20e79164d2563a9d11d2146f76f65cffb00

Its bundle requires Symfony\Component\HttpKernel\Bundle\AbstractBundle that was introduced in symfony/http-kernel:6.1

https://github.com/symfony/symfony/commit/7e8cf5df105ab0bbe0112222fa5498af486d929c

This commit fixes the build to lowest dependencies

GromNaN committed:
Add version constraint for symfony/config and zenstruck/foundry

Support for Enum cases in the configuration was added in symfony/config:6.3
https://github.com/symfony/symfony/commit/1655d1763ba210b5f7d632e4e36b1f814fd30038

Error: Object of class Zenstruck\Foundry\ORM\ResetDatabase\ResetDatabaseMode could not be converted to string

/home/runner/work/ux/ux/src/LiveComponent/vendor/symfony/config/Definition/Builder/EnumNodeDefinition.php:30
/home/runner/work/ux/ux/src/LiveComponent/vendor/zenstruck/foundry/src/ZenstruckFoundryBundle.php:152

Kocal committed:
[Toolkit][Avatar] Update templates

Kocal committed:
[TwigComponent] Minor performance improvements when using `{% props %}`

Kocal committed:
Do not automatically format our PR template

Kocal committed:
[TwigComponent] Minor performance improvement by caching `PropertyAccessor::isWritable()` calls

Kocal committed:
[Toolkit] Improve `@prop` and `@block` annotations description

Kocal committed:
[Toolkit] Rewrite `@prop` documentation to normalize `Defaults to ...`

Related to https://github.com/symfony/ux.symfony.com/issues/15

seb-jean committed:
[Toolkit] Fix `tailwind_merge` parentheses and simplify `attributes` rendering in shadcn templates

StevenRenaux committed:
Fix PHPDoc types of `TestLiveComponent::call()`

alexsegura committed:
Fix variable name in example

Kocal committed:
Fix CI

Kocal committed:
Update JavaScript dependencies

Kocal committed:
Fix unit testing script

Legacy things were forgotten, unit tests couldn't be correctly executed

Kocal committed:
Remove tsx dependency and rely on Node.js 22.18.0 native TypeScript runner

Kocal committed:
Migrate from tsup (deprecated) to tsdown

tsup is deprecated in favor of tsdown.

Follow https://github.com/symfony/ux/pull/2935, https://github.com/symfony/ux/pull/2944, and many (local) tries were I was not really happy with the files generated by tsdown/rolldown, we are finally having something extra good!

We have the benefits from https://github.com/symfony/ux/pull/2935, https://github.com/symfony/ux/pull/2944, but without their drawbacks. The code correctly follow the `es2022` target and does not do anything weird anymore with static properties (needed by controllers).

I (Claude) added a plugin to remove the region and JSDoc comments (except if they contain `@deprecated``), since I think we want to keep the code non-minified.
